src/Microsoft.CodeQuality.Analyzers/Core/ApiDesignGuidelines/StringBuilderAppendShouldNotTakeSubstring.Fixer.cs

```diff
@@ -0,0 +1,191 @@
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var typedNodeToFix =  semanticModel.GetOperation(nodeToFix) as IInvocationOperation;
+            var generator = SyntaxGenerator.GetGenerator(document);
+
+            var fixComponents = GetFixComponents(generator, typedNodeToFix);
+
+            // generate text.Length (or whatever "text" is instead in the actual code)
+            var lengthNode = generator
+                .MemberAccessExpression(
+                    fixComponents.StringArgument.Syntax, 
+                    generator.IdentifierName(nameof(string.Length)));
+
+            var startIndexNode = fixComponents.OriginalInnerArguments[0].Value.Syntax;
+
+            // generate "text".Length - 2 (where 2 is the start index given in the original code
+            var lengthArgument = generator
+                .SubtractExpression(
+                    lengthNode,
+                    startIndexNode);
+
+            // generate sb.Append(text, 2, text.Length - 2)
+            var newNode = generator.InvocationExpression(
+                fixComponents.TargetMethod, 
+                fixComponents.StringArgument.Syntax,
+                startIndexNode, 
+                lengthArgument);
+
+            // replace sb.Append(text.Substring(2)) by sb.Append(text, 2, text.Length - 2)
+            var newRoot = root
+                .ReplaceNode(
+                    nodeToFix, 
+                    newNode);
+            return document.WithSyntaxRoot(newRoot);
+        }
+
+        private struct FixComponents
+        {
+            public FixComponents(
+                IOperation stringArgument, 
+                SyntaxNode targetMethod, 
+                ImmutableArray<IArgumentOperation> originalInnerArguments)
+            {
+                StringArgument = stringArgument;
+                TargetMethod = targetMethod;
+                OriginalInnerArguments = originalInnerArguments;
+            }
+
+            public IOperation StringArgument { get; }
+            public SyntaxNode TargetMethod { get; }
+            public ImmutableArray<IArgumentOperation> OriginalInnerArguments { get; }
+        }
+
+        private static async Task<Document> FixCodeTwoParameters(
+            Document document,
+            TextSpan span,
+            CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var nodeToFix = root.FindNode(span, getInnermostNodeForTie: true);
+
+            var semanticModel = await document.GetSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            if (semanticModel.GetOperation(nodeToFix) is IInvocationOperation typedNodeToFix)
```

Comment:
this means that the user will be prompted to do something which may end up doing nothing when they try to invoke the fix.

Reply:
thanks, working on your previous comment now (register fix if and only if there's something to fix). That should make this obsolete then.

Reply:
cool.

Reply:
still checking here and still returning the original document if it fails, but as the diagnostic is registered only when the type is correct, this is just to prevent Exceptions.